src/crimson/os/seastore/journal.cc

```diff
@@ -118,35 +119,66 @@ Journal::write_record_ret Journal::write_record(
     target);
 
   auto segment_id = current_journal_segment->get_segment_id();
+  submit_queue.emplace_back(std::make_tuple(target, current_journal_segment, std::move(to_write)));
 
-  // Start write under the current exclusive stage, but wait for it
-  // in the device_submission concurrent stage to permit multiple
-  // overlapping writes.
-  auto write_fut = current_journal_segment->write(target, to_write);
   return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut
-    ).handle_error(
-      write_record_ertr::pass_further{},
-      crimson::ct_error::assert_all{
-	"Invalid error in Journal::write_record"
-      }
-    );
-  }).safe_then([this, &handle] {
+  ).then([this, &handle, target, segment_id] {
     return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target, segment_id] {
+  }).then([this, target, segment_id]() -> write_record_ret {
     logger().debug(
       "write_record: commit target {}",
       target);
     if (segment_id == current_journal_segment->get_segment_id()) {
       assert(committed_to < target);
       committed_to = target;
     }
-    return write_record_ret(
-      write_record_ertr::ready_future_marker{},
-      paddr_t{
-	segment_id,
-	target});
+    if(!write_queue.empty() || submit_queue.empty()) {
```

Comment:
I don't think write_queue actually needs to a member of Journal, it could simply be local to the lambda below.  The condition here should actually be that if the earliest entry in the submit queue is after our target, we can assume that's it's already been committed.

Reply:
"if the earliest entry in the submit queue is after our target, we can assume that's it's already been committed."  
---- I think the target of earliest entry in the submit queue should be small than current target,  the target is calculated one record by one record, so pushed into submit queue first has the smallest target, right?   so it never happen earliest entry in the submit queue is after our target.  (if I understand your meaning right).

Reply:
Sorry, I mean that if the target isn't in the queue, it's already been committed.

Reply:
if segment id is same we can compare target and first entry's target, if it is bigger than first entry's target, then it has been committed. But if cross segment, we can't use target to compare, we also can't compare if segment id > first entry's segment id, since segment id is not always monotonously increased. so I think only can check if write_queue is empty, right?

Reply:
Well, write_queue.empty() won't work with the comment I made above.  submit_record actually returns a journal_seq_t which includes an increasing sequence number for each segment.  Move the seq assignment info submit_record into write_record by modifying target to be a journal_seq_t.  modify write_record to return the journal_seq_t back to submit_record.  That would allow you to compare across segments.